### ISSUE TITLE
[Feature Request] Futureproof RetroArch with precision frame pacing presenter thread (for VRR, for BFI, for beamracing, for 120Hz/180Hz/240Hz/VRR BFI, rolling-scan CRT emulators, etc)

### ISSUE THREAD

**Issue by mdrejhon** [libretro/RetroArch#11390]
*Note: Crossposted feature suggestion from semi-related 240Hz BFI pull request at https://github.com/libretro/RetroArch/pull/11342
This was too major an ask to be inside a github pull request (, so I am creating a new thread.  To pave groundwork for improving RetroArch compatibility with improving display emulation (improving variable refresh rate to reduce lag/stytters, improving BFI with 240Hz BFI to better emulate a CRT, etc).*

This is a universal generic algorithm that should eventually become a best-practice for emulators in the next ten years.

# Problem: Existing frame pacing algorithm not future proof enough

Currently, G-SYNC uses a software-based frame pacing algorithm in RetroArch, but it is not currently optimized in a future-proof way yet:

- Problem #1: Present imprecision create visible effects (stutters during VRR, flicker during BFI) especially in CPU heavy scenarios (slower systems, intensive emulators, big RunAhead settings)
- Problem #2: Present imprecision makes it harder to add support for future enhancements (better VRR, >120Hz BFI (#10754), 
combine VRR+BFI (#10754), beam racing (#6984), rolling-scan BFI CRT emulators (#10757)

# The upgrade to existing frame pacing algorithm

I propose a **separate thread responsible for frame presenting** (e.g. Present() or glxxSwapBuffers() or whichever API) that does the following:

- **Allows frame presents to continue independently of rendering**
Presents will have no jitter even if emulator modules use darn near 100% CPU
- **Allows frame presents to be optionally ultra precise**
This can be done via busywait instead (or in addition to) of timer events -- because some algorithms (beamrace or BFI+VRR) have visibile artifacts with sub-millisecond errors.  Also, one can also timer-event to 0.5ms prior, then busywait rest of way.
- **Allows same frame pacing algorithms to work with all sync technologies**
(VSYNC ON, VSYNC OFF, DWM, triple buffer, AMD Enhanced Sync, NVIDIA Fast Sync, FreeSync, G-SYNC, VESA Adaptive-Sync, BFI, etc), making it much easier to combine them (e.g. BFI during VRR)
- **Allows easier future addition of new algorithms**
(e.g. rolling BFI CRT emulators, or lagless VSYNC beam racing), with little or no modification to emulator rendering
- **More future proof**

# Largely a Streamlining of Existing Workflow

The existing present call would be replaced by a wrapper that passes the frame to a frame presentation thread.  The presentation thread will time the presentation itself.

Some of the workflow already exists, it just needs to be re-jigged into an official unified workflow with capability of improved precision.

- "Rendering Thread" = Thread that runs the emulator and generates the emulator frames;
- "Presenting Thread" = Thread that is now permanently responsible for frame presentation;
- "Present Wrapper" = The wrapper in the Rendering Thread that transfers (or copies) the frame to the Presenting Thread

# Suggested Stage 1 Workflow

1. *Presenting Thread* only purpose is presenting frames (no rendering)
2. *Presenting Thread* is always higher priority than rendering thread, for purpose of timing precision.  Most of the time, presenting thread uses 0% CPU since it's just timing pre-rendered frames, so high precision becomes harmless to *Rendering Thread*
3. *Presenting Thread* can optionally be forced to present immediately so it is backwards compatible with existing present workflow (this can ease iterative development) or for platforms not stable with separate-thread presenting (Fortunately, I don't think there's are any left).
4. *Rendering Thread* should do all rendering, including CRT filters
5. *Present Wrapper* inside the *Rendering Thread* can be a good way to hide/centralize all the final processing.  Such as adding CRT filters, or rendering a whole sequence of BFI framebuffers (low emulator Hz on high real-display Hz).   This hides the implementation details of many refreshing algorithms, and makes cross-platform easier.
6. *Present Thread* will make sure that the time intervals between consecutive presents are as exact as possible.   When this is achieved, the algorithm suddenly become universal (works with all sync technologies).
7. *Present Wrapper* can still emulate the behavior of a 60Hz VSYNC ON waitable swapchain (regardless of whether underlying hardware is doing VSYNC ON or VSYNC OFF or VRR or BFI or whatever) by waiting for a heartbeat from the *Present Thread* 

Metaphorically, this workflow is a metaphorically software-based VSYNC ON emulator, hiding the destination's quirks/behaviors from emulator modules, while improving user-friendliness (things just works automatically upon startup) and making things less buggy (no VRR stutters, no BFI flicker).

Don't worry about BFI for now, don't worry about beamracing for now; those are solvable in future (e.g. wrappers for PresentScanLine() can be added later, as an example).  Just focus on this simple workflow.

# Easy Debugging Tip for non-VRR 60Hz Developers 

Testing without VRR can be done via 60Hz VSYNC OFF while using CPU-heavy emulation/emulation settings.  Use 60Hz VSYNC OFF, and use tearline jitter as a timing-precision debugger.  If the tearline erratically moves or jitters/vibrates massively, your present timing is not microsecond-accurate.  If the tearline is stationary or rolls slowly up/down, your present timing is near microsecond-accurate.   

1080p 60Hz is a horizontal scanrate of 67 kilohertz (approx 67000 scanlines per second, including VBI).  So a 1/67000th second delay moves a VSYNC OFF tearline downwards by 1 pixel.   Modern displays still scan from top-to-bottom ([high speed videos](https://blurbusters.com/scanout)) and VSYNC OFF tearlines are a raster artifact.   

So if your tearline is vibrating by 50 pixels up/down, that means you've got a 50/67000th second imprecision in your Present() or glxxSwapBuffers() timing.   Thusly, VSYNC OFF 60Hz is an excellent timing debugger, since VSYNC OFF tearline is a real-display raster where the new real GPU framebuffer splices into the destination display's scanout position.

If you have a high-Hz display, you can also test 60fps at VSYNC OFF 120Hz or VSYNC OFF 240Hz for more sensitive timing-precision debugging (1/135000th second for a 120Hz tearline moving downwards by 1 pixel, for example)

If you succeed, then it help guarantees VRR users get amazing framepacing, an BFI users get flicker-free operation (even if you never test VRR or BFI)  **Thus, use 60 Hz VSYNC OFF as a clever easy debugger for frame-present timing precision if you don't have 144Hz or VRR or BFI!**

**Comment by mdrejhon** [comment]
For more advanced reading about the presenter thread idea, please read the comments section of the pull request

# [https://github.com/libretro/RetroArch/pull/11342](Discussion in Variable BFI about a Presenter Thread)

There is MANY, MANY ideas there too -- and why the presenter thread does a **big universal future-proofing move for RetroArch for all non-fixed-60Hz workflows** (including VRR and BFI, plus future workflows such as beamracing).

**Comment by mdrejhon** [comment]
# Napkin Exercise Supplementary Note

## Future possible theoretical situations:

There will be cases where emulator may need to execute faster/slower than the display, so in theory, the present thread or another thread may need to provide synchronization services -- to govern the speed of the speed of emulator at a ratio higher/lower than the actual display refresh cycle itself -- for different reasons.

_(Note: For **RunAhead** workflows, the "x speed emu execute" applies **only** to the final frame of the RunAhead per emulator frame. All other rewound RunAhead frames can run at max speed in all workflows below)_

- Classical VSYNC 60 Hz Operation: fastest speed emu execute for 1/60sec; 1 hardware vsync per emu vsync
- Hardware beamrace 60Hz emu onto hardware 60Hz: 1x speed emu execute for 1/60sec; 1 hardware vsync per emu vsync
- Hardware beamrace 60Hz emu onto 120Hz: 2x speed emu execute for 1/60sec; 1 hardware vsync per emu vsync
- Hardware beamrace 60Hz emu onto 240Hz: 4x speed emu execute for 1/60sec; 1 hardware vsync per emu vsync
- Software beamrace 60Hz emu onto 240Hz: 1x speed emu execute for 1/60sec; 4 hardware vsync per emu vsync
- Software beamrace 60Hz emu onto 360Hz: 1x speed emu execute for 1/60sec; 6 hardware vsync per emu vsync
- Full screen global BFI 60Hz emu onto 120Hz: fastest emu execute for 1/60sec; 2 hardware vsync per emu vsync
- Full screen global BFI 60Hz emu onto 180Hz: fastest emu execute for 1/60sec; 4 hardware vsync per emu vsync
- Full screen global BFI 60Hz emu onto 240Hz: fastest emu execute for 1/60sec; 6 hardware vsync per emu vsync
- (VRR) Classical 60 fps Operation on >60Hz+ VRR/triplebuffer/etc: fastest speed emu hsync for 1/60sec; 1 simulated "hardware" vsync per emu vsync
- (VRR) Software beamrace 60Hz emu onto 180fps on >180Hz+ VRR: 1x speed emu hsync for 1/60sec; 3 simulated "hardware" vsync per emu vsync
- (VRR) Software beamrace 60Hz emu onto 240fps on >240Hz+ VRR: 1x speed emu hsync for 1/60sec; 4 simulated "hardware" vsync per emu vsync
- (VRR) Software beamrace 60Hz emu onto 300fps on >300Hz+ VRR: 1x speed emu hsync for 1/60sec; 5 simulated "hardware" vsync per emu vsync
- (VRR) Full screen global BFI 60Hz emu onto 180Hz on >180Hz+ VRR: fastest emu hsync for 1/60sec; 3 simulated "hardware" vsync per emu vsync
- (VRR) Full screen global BFI 60Hz emu onto 240Hz on >240Hz+ VRR: fastest emu hsync for 1/60sec; 4 simulated "hardware" vsync per emu vsync

## Glossary

- "VRR" is Variable Refresh Rate (G-SYNC, FreeSync)
- "Hardware beamrace" is the Lagless VSYNC approach (#6984)
- "Software beamrace" beam racing is the CRT electron beam emulator approach BFIv3 (#10757)

**Comment by mdrejhon** [comment]
# TL;DR Version Of This Feature Request

1. RetroArch frame-presenter (swapchain) thread made separate from emulator/renderer thread 
2. Frame presenter thread (thread responsible for swapchain) is higher priority than renderer thread.

This will improve reliability even further and add some magical powers.
